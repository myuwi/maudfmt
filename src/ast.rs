use crate::token::{Token, TokenWithTrivia};

#[derive(Debug)]
pub struct Markup {
    pub root: Block,
}

#[derive(Debug)]
pub struct Block {
    pub open_brace: TokenWithTrivia,
    pub nodes: Vec<Node>,
    pub close_brace: TokenWithTrivia,
}

#[derive(Debug)]
pub enum Node {
    Element(Element),
    Block(Block),
    Str(Str),
}

impl Node {
    pub fn surrounding_trivia(&self) -> (&Vec<Token>, &Vec<Token>) {
        match self {
            Node::Element(element) => {
                let leading_trivia = &element.tag.leading_trivia;
                let trailing_trivia = match &element.body {
                    ElementBody::Block(block) => &block.close_brace.trailing_trivia,
                    ElementBody::Void(semi) => &semi.trailing_trivia,
                };

                (leading_trivia, trailing_trivia)
            }
            Node::Block(block) => {
                let leading_trivia = &block.open_brace.leading_trivia;
                let trailing_trivia = &block.close_brace.trailing_trivia;

                (leading_trivia, trailing_trivia)
            }
            Node::Str(str) => {
                let leading_trivia = &str.0.leading_trivia;
                let trailing_trivia = &str.0.trailing_trivia;

                (leading_trivia, trailing_trivia)
            }
        }
    }
}

#[derive(Debug)]
pub struct Element {
    pub tag: TokenWithTrivia,
    pub attrs: Vec<Attribute>,
    pub body: ElementBody,
}

#[derive(Debug)]
pub enum ElementBody {
    Block(Block),
    Void(TokenWithTrivia),
}

#[derive(Debug)]
pub struct Attribute {
    pub name: TokenWithTrivia,
    pub eq: TokenWithTrivia,
    pub value: TokenWithTrivia,
}

#[derive(Debug)]
pub struct Str(pub TokenWithTrivia);
