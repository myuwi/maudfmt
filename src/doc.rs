use pretty::RcDoc;

use crate::{
    ast::{Block, Element, ElementBody, Markup, Node, Str},
    kind::TokenKind,
    token::Token,
};

const INDENT: isize = 4;

pub trait Doc {
    fn to_doc(&self) -> RcDoc;
}

impl Doc for Markup {
    fn to_doc(&self) -> RcDoc {
        self.root.to_doc()
    }
}

impl Doc for Block {
    fn to_doc(&self) -> RcDoc {
        let mut doc = RcDoc::nil()
            .append(self.open_brace.token.text())
            .append(handle_trailing_trivia(&self.open_brace.trailing_trivia))
            .append(self.nodes.to_doc())
            .append(handle_leading_trivia(
                &self.close_brace.leading_trivia,
                true,
            ))
            .nest(INDENT)
            .append(self.close_brace.token.text());

        let prefer_fold = !self
            .open_brace
            .trailing_trivia
            .iter()
            .any(|t| t.kind == TokenKind::Newline);

        if prefer_fold {
            doc = doc.group();
        }

        doc
    }
}

// TODO: handle blank lines between nodes
impl Doc for Vec<Node> {
    fn to_doc(&self) -> RcDoc {
        RcDoc::concat(self.iter().map(|node| node.to_doc()))
    }
}

impl Doc for Node {
    fn to_doc(&self) -> RcDoc {
        let (leading_trivia, trailing_trivia) = self.surrounding_trivia();

        let doc = handle_leading_trivia(leading_trivia, false);

        let doc = doc.append(match self {
            Node::Element(element) => element.to_doc(),
            Node::Block(block) => block.to_doc(),
            Node::Str(str) => str.to_doc(),
        });

        let doc = doc.append(handle_trailing_trivia(trailing_trivia));

        doc
    }
}

impl Doc for Element {
    fn to_doc(&self) -> RcDoc {
        let doc = RcDoc::text(self.tag.token.text())
            .append(handle_trailing_trivia(&self.tag.trailing_trivia))
            .group();

        let doc = doc.append(self.body.to_doc());

        doc
    }
}

impl Doc for ElementBody {
    fn to_doc(&self) -> RcDoc {
        match self {
            ElementBody::Block(block) => {
                let doc = handle_leading_trivia(&block.open_brace.leading_trivia, false);
                doc.append(block.to_doc())
            }
        }
    }
}

impl Doc for Str {
    fn to_doc(&self) -> RcDoc {
        RcDoc::text(self.0.token.text())
    }
}

fn handle_leading_trivia(leading_trivia: &[Token], ending_newline: bool) -> RcDoc {
    let mut doc = RcDoc::nil();

    let mut trivia_iter = leading_trivia
        .iter()
        .filter(|t| t.kind != TokenKind::Whitespace)
        .peekable();

    while let Some(token) = trivia_iter.next() {
        let comment_doc = match token.kind {
            TokenKind::LineComment => RcDoc::text(token.text()).append(RcDoc::hardline()),
            TokenKind::BlockComment => {
                let text = RcDoc::text(token.text());

                let line = match trivia_iter.peek().map(|t| t.kind) {
                    Some(TokenKind::Newline) | None if ending_newline => RcDoc::line(),
                    _ => RcDoc::softline(),
                };

                text.append(line)
            }
            _ => continue,
        };

        doc = doc.append(comment_doc);
    }

    doc
}

fn handle_trailing_trivia(trailing_trivia: &[Token]) -> RcDoc {
    let trailing_comments = trailing_trivia
        .iter()
        .filter(|t| t.kind.is_comment())
        .collect::<Vec<_>>();

    let doc = RcDoc::concat(
        trailing_comments
            .iter()
            .flat_map(|t| vec![RcDoc::space(), RcDoc::text(t.text())]),
    );

    let has_line_comment = trailing_comments
        .iter()
        .any(|t| t.kind == TokenKind::LineComment);

    let doc = doc.append(if has_line_comment {
        RcDoc::hardline()
    } else {
        RcDoc::line()
    });

    doc
}
