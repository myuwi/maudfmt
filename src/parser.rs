use std::process::exit;

use miette::{miette, LabeledSpan, Severity};
use nom::{
    branch::alt,
    bytes::complete::{escaped, tag},
    character::complete::{alpha1, alphanumeric1, char, multispace0, multispace1, none_of, one_of},
    combinator::{cut, map, recognize, value},
    error::VerboseError,
    multi::{many0, separated_list0},
    sequence::{delimited, pair, preceded, separated_pair, terminated, tuple},
    IResult,
};

// TODO: Raw strings
// TODO: Classes and IDs: .foo #bar
// TODO: Implicit div elements
// TODO: Splices
// TODO: Toggles
// TODO: Control structures

#[derive(Clone, Debug)]
pub struct Block {
    pub newline: bool,
    pub markup: Markup,
}

#[derive(Clone, Debug)]
pub enum ElementBody {
    Void,
    Block(Block),
}

#[derive(Clone, Debug)]
pub enum AttributeValue {
    String(String),
    Empty,
}

#[derive(Clone, Debug)]
pub struct Attribute {
    pub name: String,
    pub value: AttributeValue,
}

#[derive(Clone, Debug)]
pub struct Element {
    pub name: String,
    pub attrs: Vec<Attribute>,
    pub body: ElementBody,
}

#[derive(Clone, Debug)]
pub enum Node {
    Element(Element),
    Block(Block),
    Str(String),
}

#[derive(Clone, Debug)]
pub struct Markup(pub Vec<Node>);

fn string(input: &str) -> IResult<&str, String, VerboseError<&str>> {
    map(
        preceded(
            char('"'),
            terminated(escaped(none_of(r#"\""#), '\\', one_of(r#""n\"#)), char('"')),
        ),
        String::from,
    )(input)
}

fn block(input: &str) -> IResult<&str, Block, VerboseError<&str>> {
    delimited(
        preceded(multispace0, char('{')),
        map(tuple((multispace0, markup)), |(whitespace, markup)| Block {
            newline: whitespace.contains('\n'),
            markup,
        }),
        preceded(multispace0, cut(char('}'))),
    )(input)
}

fn void(input: &str) -> IResult<&str, (), VerboseError<&str>> {
    value((), preceded(multispace0, char(';')))(input)
}

fn body(input: &str) -> IResult<&str, ElementBody, VerboseError<&str>> {
    alt((
        value(ElementBody::Void, void),
        map(block, ElementBody::Block),
    ))(input)
}

fn non_empty_attribute(input: &str) -> IResult<&str, Attribute, VerboseError<&str>> {
    map(separated_pair(tag_name, char('='), string), |a| Attribute {
        name: a.0.to_string(),
        value: AttributeValue::String(a.1),
    })(input)
}

fn empty_attribute(input: &str) -> IResult<&str, Attribute, VerboseError<&str>> {
    map(tag_name, |a| Attribute {
        name: a.to_string(),
        value: AttributeValue::Empty,
    })(input)
}

fn attrs(input: &str) -> IResult<&str, Vec<Attribute>, VerboseError<&str>> {
    separated_list0(multispace1, alt((non_empty_attribute, empty_attribute)))(input)
}

fn tag_name(input: &str) -> IResult<&str, &str, VerboseError<&str>> {
    recognize(pair(
        alpha1,
        many0(alt((alphanumeric1, tag("_"), tag("-")))),
    ))(input)
}

fn element(input: &str) -> IResult<&str, Element, VerboseError<&str>> {
    map(
        tuple((
            tag_name,
            preceded(multispace0, attrs),
            preceded(multispace0, body),
        )),
        |(name, attrs, body)| Element {
            name: name.to_string(),
            attrs,
            body,
        },
    )(input)
}

fn markup(input: &str) -> IResult<&str, Markup, VerboseError<&str>> {
    map(
        many0(preceded(
            multispace0,
            alt((
                map(element, Node::Element),
                map(string, Node::Str),
                map(block, Node::Block),
            )),
        )),
        Markup,
    )(input)
}

pub fn parse(src: &str, full_file: &str) -> Markup {
    match markup(src) {
        Ok((_, markup)) => markup,
        Err(e) => {
            let e = match e {
                nom::Err::Error(e) | nom::Err::Failure(e) => e,
                nom::Err::Incomplete(_) => {
                    eprintln!("Incomplete input");
                    exit(1);
                }
            };

            let context = e.errors.first().unwrap();
            let offset = full_file.find(context.0).unwrap_or_default();

            let text = "Unknown token";

            let report = miette!(
                severity = Severity::Error,
                labels = vec![LabeledSpan::at_offset(offset, text)],
                "Parsing error"
            )
            .with_source_code(full_file.to_string());

            eprintln!("{:?}", report);
            exit(1)
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_block_element() {
        let expected = r#"Ok(("", Element { name: "h1", attrs: [], body: Block(Block { newline: false, markup: Markup([Str("Poem")]) }) }))"#;

        assert_eq!(format!("{:?}", element(r#"h1 { "Poem" }"#)), expected);
        assert_eq!(format!("{:?}", element(r#"h1{"Poem"}"#)), expected);
        assert_eq!(
            format!(
                "{:?}",
                element(
                    r#"h1
                    {"Poem"
                    }"#
                )
            ),
            expected
        );
    }

    #[test]
    fn test_parse_void_element() {
        assert_eq!(
            format!("{:?}", element(r#"input;"#)),
            r#"Ok(("", Element { name: "input", attrs: [], body: Void }))"#
        );
    }

    #[test]
    fn test_parse_attrs() {
        assert_eq!(
            format!("{:?}", element(r#"input type="checkbox" checked;"#)),
            r#"Ok(("", Element { name: "input", attrs: [Attribute { name: "type", value: String("checkbox") }, Attribute { name: "checked", value: Empty }], body: Void }))"#
        );
    }

    #[test]
    fn test_parse_block() {
        assert_eq!(
            format!("{:?}", markup(r#"{ input; }"#)),
            r#"Ok(("", Markup([Block(Block { newline: false, markup: Markup([Element(Element { name: "input", attrs: [], body: Void })]) })])))"#
        );

        assert_eq!(
            format!(
                "{:?}",
                block(
                    r#"{
                    input;
                }"#
                )
            ),
            r#"Ok(("", Block { newline: true, markup: Markup([Element(Element { name: "input", attrs: [], body: Void })]) }))"#
        );
    }

    #[test]
    fn test_parse_multiroot() {
        assert_eq!(
            format!("{:?}", markup(r#""a" "b""#)),
            r#"Ok(("", Markup([Str("a"), Str("b")])))"#
        );
    }
}
