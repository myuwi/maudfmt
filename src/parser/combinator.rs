use nom::{character::complete::multispace0, sequence::preceded};

pub fn ws<'a, O, E, F>(f: F) -> impl nom::Parser<&'a str, O, E>
where
    E: nom::error::ParseError<&'a str>,
    F: nom::Parser<&'a str, O, E>,
{
    preceded(multispace0, f)
}
