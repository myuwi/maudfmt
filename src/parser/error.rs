use nom::error::ErrorKind;

#[derive(Clone, Debug, PartialEq)]
pub struct ParserError<I> {
    pub errors: Vec<(I, ParserErrorKind)>,
}

#[derive(Clone, Debug, PartialEq)]
pub enum ParserErrorKind {
    Char(char),
    Nom(ErrorKind),
}

pub trait Offset {
    fn offset(&self, other: &Self) -> isize;
}

impl Offset for &str {
    fn offset(&self, other: &Self) -> isize {
        self.as_ptr() as isize - other.as_ptr() as isize
    }
}

impl<I> nom::error::ParseError<I> for ParserError<I>
where
    I: Offset,
{
    fn from_error_kind(input: I, kind: nom::error::ErrorKind) -> Self {
        ParserError {
            errors: vec![(input, ParserErrorKind::Nom(kind))],
        }
    }

    fn append(input: I, kind: nom::error::ErrorKind, mut other: Self) -> Self {
        other.errors.push((input, ParserErrorKind::Nom(kind)));
        other
    }

    fn from_char(input: I, c: char) -> Self {
        ParserError {
            errors: vec![(input, ParserErrorKind::Char(c))],
        }
    }

    fn or(self, other: Self) -> Self {
        let self_first = self.errors.first();
        let other_first = other.errors.first();

        match (self_first, other_first) {
            (_, None) => self,
            (None, _) => other,
            (Some(s), Some(o)) => {
                let diff = s.0.offset(&o.0);

                if diff > 0 {
                    self
                } else {
                    other
                }
            }
        }
    }
}
