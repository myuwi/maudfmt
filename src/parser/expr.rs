use nom::{
    branch::alt,
    bytes::complete::tag,
    character::complete::{char, none_of},
    combinator::{not, peek, recognize},
    multi::many0_count,
    sequence::{delimited, terminated},
    InputTake,
};

use super::{
    combinator::cond_err,
    literal::{char_lit, str_lit},
    NomResult,
};

pub fn group<'a>(start_delim: char, end_delim: char) -> impl FnMut(&'a str) -> NomResult<&str> {
    move |i| {
        recognize(delimited(
            char(start_delim),
            expr_impl(true, true),
            char(end_delim),
        ))(i)
    }
}

// TODO: handle comments
fn expr_impl<'a>(eager_brace: bool, nested_expr: bool) -> impl FnMut(&'a str) -> NomResult<&str> {
    move |i| {
        recognize(many0_count(alt((
            str_lit,
            char_lit,
            group('(', ')'),
            cond_err(eager_brace, group('{', '}')),
            group('[', ']'),
            // TODO: probably not the optimal way...
            recognize(none_of("(){}[];=")),
            terminated(tag("="), not(peek(tag(">")))),
            cond_err(nested_expr, alt((tag(";"), tag("=>")))),
        ))))(i)
    }
}

pub fn expr<'a>(eager_brace: bool) -> impl FnMut(&'a str) -> NomResult<&str> {
    move |i| expr_impl(eager_brace, false)(i).map(|(_, o)| i.take_split(o.trim_end().len()))
}
