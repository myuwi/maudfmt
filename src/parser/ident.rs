use nom::{
    bytes::complete::{tag, take_while, take_while1},
    combinator::{map_parser, recognize},
    sequence::pair,
    AsChar,
};

use super::NomResult;

pub fn keyword<'a>(kw: &'a str) -> impl FnMut(&'a str) -> NomResult<&str> {
    map_parser(identifier, tag(kw))
}

pub fn identifier(input: &str) -> NomResult<&str> {
    recognize(pair(
        take_while1(|c: char| c.is_alpha() || c == '_' || c >= '\u{0080}'),
        take_while(|c: char| c.is_alphanum() || c == '_' || c >= '\u{0080}'),
    ))(input)
}
