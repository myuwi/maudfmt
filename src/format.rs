use pretty::RcDoc;

use crate::ast::Markup;
use crate::doc::Doc;

pub fn pretty_print(markup: &Markup, indent: usize, width: usize) -> String {
    let doc = markup.to_doc();
    let doc = RcDoc::text(" ".repeat(indent))
        .append(doc)
        .nest(indent as isize);

    doc.pretty(width).to_string()
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::parser::Parser;

    #[test]
    fn basic() {
        let input = r#"{
h1 {         "Hello "
   b {
       "world"}
}p {
"Lorem ipsum"
}
        }"#;

        let markup = Parser::new(input, 0).parse().unwrap();
        let formatted = pretty_print(&markup, 0, 100);

        insta::assert_snapshot!(formatted);
    }

    #[test]
    fn too_many_comments() {
        let input = r#"{ // line 1
// line 2
/* block 1 */h1 /* block 2 */ // line 3
            /* block 3 */ {  // line 4
                /* block 4 */ "Hello world" // line 5
            /* block 5 */            /* block 6 */
            // line 6
            } // line 7
        /* block 7 */        /* block 8 */
        /* block 9 */ }"#;

        let markup = Parser::new(input, 0).parse().unwrap();
        let formatted = pretty_print(&markup, 0, 100);

        insta::assert_snapshot!(formatted);
    }

    #[test]
    fn comments_fold() {
        let input = r#"{ /* 1 */ h1
            /* 2 */
             {          
                 /* 3 */
                 "Hello world" /* 4 */
                 /* 5 */
            }
            /* 6 */
        }"#;

        let markup = Parser::new(input, 0).parse().unwrap();
        let formatted = pretty_print(&markup, 0, 100);

        insta::assert_snapshot!(formatted);
    }
}
