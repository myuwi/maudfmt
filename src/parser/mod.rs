use std::ops::Range;

use miette::SourceSpan;
use nom::{
    branch::alt,
    bytes::complete::tag,
    character::complete::{alpha1, alphanumeric1, char, multispace0, multispace1},
    combinator::{all_consuming, cut, fail, map, opt, recognize, value},
    error::VerboseError,
    multi::{many0, separated_list0},
    sequence::{delimited, pair, preceded, separated_pair, tuple},
    Finish,
};

mod combinator;
mod expr;
mod ident;
mod literal;
mod path;
mod pattern;

use combinator::ws;
use expr::expr;
use ident::keyword;
use literal::str_lit;
use pattern::pattern;

use crate::error::ParseError;

pub type NomResult<'a, O> = Result<(&'a str, O), nom::Err<VerboseError<&'a str>>>;

#[derive(Clone, Debug)]
pub struct For<'a> {
    pub pattern: &'a str,
    pub expr: &'a str,
    pub body: Block<'a>,
}

#[derive(Clone, Debug)]
pub enum Else<'a> {
    If(If<'a>),
    Then(Block<'a>),
}

#[derive(Clone, Debug)]
pub struct If<'a> {
    pub cond: &'a str,
    pub body: Block<'a>,
    pub else_clause: Option<Box<Else<'a>>>,
}

#[derive(Clone, Debug)]
pub enum ControlStructure<'a> {
    If(If<'a>),
    For(For<'a>),
}

#[derive(Clone, Debug)]
pub struct Splice<'a> {
    pub expr: &'a str,
}

#[derive(Clone, Debug)]
pub struct Block<'a> {
    pub newline: bool,
    pub nodes: Vec<Node<'a>>,
}

#[derive(Clone, Debug)]
pub enum ElementBody<'a> {
    Void,
    Block(Block<'a>),
}

#[derive(Clone, Debug)]
pub enum AttributeValue<'a> {
    String(&'a str),
    Empty,
}

#[derive(Clone, Debug)]
pub struct Attribute<'a> {
    pub name: &'a str,
    pub value: AttributeValue<'a>,
}

#[derive(Clone, Debug)]
pub struct Element<'a> {
    pub name: &'a str,
    pub attrs: Vec<Attribute<'a>>,
    pub body: ElementBody<'a>,
}

#[derive(Clone, Debug)]
pub enum Node<'a> {
    Element(Element<'a>),
    Block(Block<'a>),
    StrLit(&'a str),
    Splice(Splice<'a>),
    ControlStructure(ControlStructure<'a>),
}

#[derive(Clone, Debug)]
pub struct Markup<'a> {
    pub nodes: Vec<Node<'a>>,
}

fn for_expr(input: &str) -> NomResult<For> {
    preceded(
        keyword("for"),
        map(
            tuple((ws(pattern), keyword("in"), ws(expr(false)), block)),
            |(pattern, _, expr, body)| For {
                pattern,
                expr,
                body,
            },
        ),
    )(input)
}

fn else_expr(input: &str) -> NomResult<Else> {
    preceded(
        keyword("else"),
        ws(alt((map(if_expr, Else::If), map(block, Else::Then)))),
    )(input)
}

fn if_expr(input: &str) -> NomResult<If> {
    let opt_else = opt(preceded(char('@'), map(else_expr, Box::new)));

    preceded(
        keyword("if"),
        map(
            tuple((ws(expr(false)), block, ws(opt_else))),
            |(cond, body, else_clause)| If {
                cond,
                body,
                else_clause,
            },
        ),
    )(input)
}

fn control_structure(input: &str) -> NomResult<ControlStructure> {
    preceded(
        char('@'),
        alt((
            map(if_expr, ControlStructure::If),
            map(for_expr, ControlStructure::For),
            cut(fail),
        )),
    )(input)
}

fn splice(input: &str) -> NomResult<Splice> {
    map(delimited(char('('), ws(expr(true)), char(')')), |expr| {
        Splice { expr }
    })(input)
}

fn block(input: &str) -> NomResult<Block> {
    delimited(
        char('{'),
        map(pair(multispace0, ws(nodes)), |(whitespace, nodes)| Block {
            newline: whitespace.contains('\n'),
            nodes,
        }),
        cut(char('}')),
    )(input)
}

fn void(input: &str) -> NomResult<()> {
    value((), char(';'))(input)
}

fn body(input: &str) -> NomResult<ElementBody> {
    alt((
        value(ElementBody::Void, void),
        map(block, ElementBody::Block),
        cut(fail),
    ))(input)
}

fn non_empty_attribute(input: &str) -> NomResult<Attribute> {
    map(separated_pair(tag_name, char('='), str_lit), |a| {
        Attribute {
            name: a.0,
            value: AttributeValue::String(a.1),
        }
    })(input)
}

fn empty_attribute(input: &str) -> NomResult<Attribute> {
    map(tag_name, |a| Attribute {
        name: a,
        value: AttributeValue::Empty,
    })(input)
}

fn attrs(input: &str) -> NomResult<Vec<Attribute>> {
    separated_list0(multispace1, alt((non_empty_attribute, empty_attribute)))(input)
}

fn tag_name(input: &str) -> NomResult<&str> {
    recognize(pair(
        alpha1,
        many0(alt((alphanumeric1, tag("_"), tag("-")))),
    ))(input)
}

fn element(input: &str) -> NomResult<Element> {
    map(tuple((tag_name, ws(attrs), body)), |(name, attrs, body)| {
        Element { name, attrs, body }
    })(input)
}

fn nodes(input: &str) -> NomResult<Vec<Node>> {
    many0(preceded(
        multispace0,
        alt((
            map(element, Node::Element),
            map(str_lit, Node::StrLit),
            map(block, Node::Block),
            map(splice, Node::Splice),
            map(control_structure, Node::ControlStructure),
        )),
    ))(input)
}

fn markup(input: &str) -> Result<Markup, VerboseError<&str>> {
    all_consuming(map(nodes, |n| Markup { nodes: n }))(input)
        .finish()
        .map(|(_, markup)| markup)
}

pub fn parse_range(src: &str, range: Range<usize>) -> Result<Markup, ParseError> {
    let content = src[range].trim();

    markup(content).map_err(|e| {
        dbg!(&e);
        let (remaining_input, _) = e.errors.first().unwrap();
        let offset = src.find(remaining_input).unwrap_or_default();

        ParseError::UnexpectedToken {
            src: src.to_string(),
            err_span: SourceSpan::from((offset, 0)),
        }
    })
}
