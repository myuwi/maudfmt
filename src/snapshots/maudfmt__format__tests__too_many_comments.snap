---
source: src/format.rs
expression: formatted
---
{ // line 1
    // line 2
    /* block 1 */ h1 /* block 2 */ // line 3
    /* block 3 */ { // line 4
        /* block 4 */ "Hello world" // line 5
        /* block 5 */ /* block 6 */
        // line 6
    } // line 7
    /* block 7 */ /* block 8 */
    /* block 9 */
}
